eos-updater/poll-common.c

```diff
@@ -1186,6 +1186,69 @@ string_to_download_source (const gchar *str,
   return TRUE;
 }
 
+static gboolean
+get_commit_sizes (OstreeRepo    *repo,
+                  const gchar   *checksum,
+                  gint64        *new_archived,
+                  gint64        *new_unpacked,
+                  gint64        *archived,
+                  gint64        *unpacked,
+                  GCancellable  *cancellable,
+                  GError       **error)
+{
+  g_autoptr(GVariant) commit = NULL;
+  g_autoptr(GPtrArray) sizes = NULL;
+
+  g_return_val_if_fail (new_archived != NULL, FALSE);
+  g_return_val_if_fail (new_unpacked != NULL, FALSE);
+  g_return_val_if_fail (archived != NULL, FALSE);
+  g_return_val_if_fail (unpacked != NULL, FALSE);
+
+#ifdef HAVE_OSTREE_COMMIT_GET_OBJECT_SIZES
+  if (!ostree_repo_load_variant (repo, OSTREE_OBJECT_TYPE_COMMIT, checksum,
+                                 &commit, error))
+    {
+      g_prefix_error (error, "Failed to read commit: ");
+      return FALSE;
+    }
+
+  if (!ostree_commit_get_object_sizes (commit, &sizes, error))
+    return FALSE;
+
+  for (guint i = 0; i < sizes->len; i++)
+    {
+      OstreeCommitSizesEntry *entry = sizes->pdata[i];
+      gboolean exists;
+
+      *archived += (gint64) entry->archived;
+      *unpacked += (gint64) entry->unpacked;
+
+      if (!ostree_repo_has_object (repo, entry->objtype, entry->checksum,
+                                   &exists, cancellable, error))
+        return FALSE;
+
+      if (!exists)
+        {
+          /* Object not in local repo */
+          *new_archived += (gint64) entry->archived;
+          *new_unpacked += (gint64) entry->unpacked;
```

Comment:
We should probably use saturated addition here. I’d also be tempted to use `guint64` for the types in `get_commit_sizes()` and then clamp down to `gint64` in the calling function, to give that function the scope for how it handles the possibility of sizes which are greater than `G_MAXINT64`.